pydantic_ai_slim/pydantic_ai/toolsets/prepared.py

```diff
@@ -34,3 +38,115 @@ async def get_tools(self, ctx: RunContext[AgentDepsT]) -> dict[str, ToolsetTool[
             name: replace(original_tools[name], tool_def=tool_def)
             for name, tool_def in prepared_tool_defs_by_name.items()
         }
+
+
+@dataclass
+class ToolConfigPreparedToolset(WrapperToolset[AgentDepsT]):
+    """A toolset that prepares the tools it contains using a ToolConfig."""
+
+    tool_config: dict[str, ToolConfig]
+
+    async def get_tools(self, ctx: RunContext[AgentDepsT]) -> dict[str, ToolsetTool[AgentDepsT]]:
+        original_tools = await super().get_tools(ctx)
+
+        # Start with a shallow copy to avoid mutating the parent's dict
+        result_tools = dict(original_tools)
+
+        # Iterate tool_config - skip tools that don't exist in this toolset
+        # (tool_config may be shared across multiple toolsets, e.g. function tools + output tools)
+        for tool_name, config in self.tool_config.items():
+            if tool_name not in original_tools:
+                continue
+
+            tool = original_tools[tool_name]
+            original_tool_def = tool.tool_def
+            parameters_json_schema = copy.deepcopy(original_tool_def.parameters_json_schema)
+
+            if config.tool_args_descriptions:
+                self._update_arg_descriptions(parameters_json_schema, config.tool_args_descriptions, tool_name)
+
+            updated_tool_def = replace(
+                original_tool_def,
+                parameters_json_schema=parameters_json_schema,
+                **{
+                    k: v
+                    for k, v in {
+                        'name': config.name,
+                        'description': config.tool_description,
+                        'strict': config.strict,
+                    }.items()
+                    if v is not None
+                },
+            )
+
+            updated_tool = replace(tool, tool_def=updated_tool_def)
+
+            # Handle renaming: remove old key if renamed, then add with final name
+            final_tool_name = config.name if config.name is not None else tool_name
```

Comment:
To keep the renamed tools working when they are called with the new names, we need to do mapping in `call_tool` as well. Check out `RenamedToolset`. 

I think we should be able to implement this toolset as a combination of `RenamedToolset` + `PreparedToolset`, both with custom functions.